source/postprocess/visualization/grain_lag_angle.cc

```diff
@@ -0,0 +1,178 @@
+                // dF/dt = LF
+                // where L is the velocity gradient tensor, and F(t=0) = I. The solution to this is
+                // F = exp(Lt)
+                // which can be written as an infinite sum:
+                // F = I + Lt + L^2*t^2/2! + L^3*t^3/3! + ...
+                // and then the "left stretch" tensor is U = F * F^T
+                // We truncate the series sum at the L^4 term. Note that L is sometimes
+                // taken to be the strain rate tensor (ie the velocity gradient without
+                // the rotational component).
+                // The eigenvectors of U are the directions of the major axes of the finite
+                // strain ellipsoid, so for the *infinite* strain axis, we want the eigenvector
+                // corresponding to the largest eigenvalue of U in the limit as t->infinity.
+                // As per Kaminski and Ribe (2002), tmax = 75*tauISA is infinite enough for
+                // most applications.
+                unsigned int maxorder = 4; // truncate at L^4
+                Tensor<2, dim> F = unit_symmetric_tensor<dim, double>(); // initialize as identity (first term of sum)
+                const double tmax = 75 * tauISA; // "infinite" time
+
+                for (unsigned int i = 1; i <= maxorder; ++i) // loop and calculate terms of sum
+                  {
+                    Tensor<2, dim> multiplier = unit_symmetric_tensor<dim, double>();  // to collect powers of L
+                    double factor = 1.0;  // to collect powers of tmax and factorials
+                    for (unsigned int j = 0; j < i; ++j)
+                      {
+                        multiplier = multiplier
+                                     * Tensor<2, dim>(velocity_gradient[0]);
+                        factor *= tmax / (j + 1);
+                      }
+                    multiplier = factor * multiplier;
+                    F += multiplier;  // sum terms
+                  }
+
+                // The "left stretch" tensor is computed from the deformation gradient tensor
+                // Note that Kaminski and Ribe incorrectly wrote U=Ft*F (typo, I guess).
+                const Tensor<2, dim> Ft = transpose(F);
+                const SymmetricTensor<2, dim> U = SymmetricTensor<2, dim>(F * Ft);
+                // The ~infinite strain axis is the first eigenvector of the "left stretch" tensor.
+                const Tensor<1, dim> ehat = eigenvectors(U)[0].second;
+
+                // Calculate theta by dotting the ISA (ehat) and the velocity (scaled to a unit vector)
+                // theta is limited to be less than pi/2 (stay in the first quadrant)
+                const double umag = std::sqrt(velocities[0] * velocities[0]);
+                const double umagmin = 1.0e-40;  // to avoid problems with /0 in acos
+                double theta_val = 0;
+                if (umag > umagmin)
```

Comment:
there are libraries that make it unnecessary to use small arbitrary numbers. I would suggest rewriting this as:
```
if (umag > 100.0 * std::numeric_limits<double>::min())
```
This way it will be tied to the precision of the system you are using.